File: LIB/toolbox.f90
MODULE toolbox
IMPLICIT NONE
CONTAINS
!=======================================================================
! PURPOSE: increment and add a reference to the reference list provided
! as input.
!=======================================================================
SUBROUTINE addref(progname,reference,nref,reflist,chem)
  USE keyparameter, ONLY: mxlfo
  IMPLICIT NONE
  CHARACTER(LEN=*),INTENT(IN) :: progname
  CHARACTER(LEN=*),INTENT(IN) :: reference
  INTEGER,INTENT(INOUT) :: nref
  CHARACTER(LEN=*),INTENT(INOUT) :: reflist(:)
  CHARACTER(LEN=*),INTENT(IN),OPTIONAL :: chem

  CHARACTER(LEN=mxlfo) :: species
  INTEGER              :: i

! check if chem is provided
  IF (PRESENT(chem)) THEN; species=chem ; ELSE ; species=' ' ; ENDIF

  DO i=1,nref
    IF (reflist(i)==reference) RETURN ! ref already in list
  ENDDO

  nref=nref+1
  IF (nref> SIZE(reflist)) THEN
    PRINT*, "--error-- too many references added for the list"
    PRINT*, "          in subroutine:",TRIM(progname)
    PRINT*, "size of reflist is: ", SIZE(reflist)
    IF (species(1:1)/=' ') &
    PRINT*, "          for the species:",TRIM(species)
    PRINT*, "          reference was:",TRIM(reference)
    PRINT*,reflist(:)
    STOP "in addref"  
  ENDIF
  reflist(nref)=reference
END SUBROUTINE addref
  
!=======================================================================
! PURPOSE: increment number of reactions.
!=======================================================================
SUBROUTINE addrx(progname,chem,nr,flag)
  IMPLICIT NONE

  CHARACTER(LEN=*),INTENT(IN) :: progname
  CHARACTER(LEN=*),INTENT(IN) :: chem
  INTEGER,INTENT(INOUT) :: nr
  INTEGER,INTENT(INOUT) :: flag(:)
  CHARACTER(LEN=70)     :: mesg

  nr=nr+1
  IF (nr>SIZE(flag)) THEN
    mesg='-error- too many reactions created for species: '//TRIM(chem)//' in addrx'
    CALL stoperr(progname,mesg,chem)
!    PRINT*, "--error-- too many reactions created for species: ", TRIM(chem)
!    PRINT*, "          in subroutine: ",TRIM(progname)
!    STOP "in addrx"
  ENDIF
  flag(nr) = 1
END SUBROUTINE addrx

!=======================================================================
! PURPOSE: increment number of products in a reaction.
!=======================================================================
SUBROUTINE add1tonp(progname,chem,np)
  USE keyparameter, ONLY: mxpd  ! max # of possible products in a reaction
  IMPLICIT NONE

  CHARACTER(LEN=*),INTENT(IN) :: progname
  CHARACTER(LEN=*),INTENT(IN) :: chem
  INTEGER,INTENT(INOUT) :: np

  np=np+1
  IF (np>mxpd) THEN
    PRINT*, "--error-- too many products created, np > mxpd"
    PRINT*, "          for species: ", TRIM(chem)
    PRINT*, "          in subroutine: ",TRIM(progname)
    STOP "in add1tonp"
  ENDIF
END SUBROUTINE add1tonp

!=======================================================================
! PURPOSE: generates error message output
!=======================================================================
SUBROUTINE stoperr(prog,mesg,chem)
  IMPLICIT NONE
  CHARACTER(LEN=*),INTENT(IN) :: chem  ! formula of species triggering error
  CHARACTER(LEN=*),INTENT(IN) :: prog  ! name of calling prog routine
  CHARACTER(LEN=*),INTENT(IN) :: mesg  ! nature of problem

!  LOGICAL,PARAMETER :: kill=.FALSE.
  LOGICAL,PARAMETER :: kill=.TRUE.

  WRITE(6,'(a)') ' '
  WRITE(6,'(a)') '--error-- in: '//TRIM(prog)
  WRITE(6,'(a)') TRIM(mesg)
  WRITE(6,'(a)') 'for species: '//TRIM(chem)
  IF (kill) STOP "in stoperr"

END SUBROUTINE stoperr

!=======================================================================
! PURPOSE: set bond value
!=======================================================================
SUBROUTINE setbond(bondtb,x,y,bondvalue)
  IMPLICIT NONE

  INTEGER,INTENT(IN) :: x,y
  INTEGER,INTENT(IN) :: bondvalue
  INTEGER,INTENT(INOUT) :: bondtb(:,:)

  bondtb(x,y)=bondvalue
  bondtb(y,x)=bondvalue

END SUBROUTINE setbond

!===========================================================
! PURPOSE: remove empty nodes in group and bond
!===========================================================
SUBROUTINE erase_blank(bond,group)
  IMPLICIT NONE

  INTEGER,INTENT(INOUT)          :: bond(:,:)
  CHARACTER(LEN=*),INTENT(INOUT) :: group(:)

  INTEGER         :: i,j,k,last,n,nca,icheck
  INTEGER         :: tbond(SIZE(bond,1),SIZE(bond,2))
  CHARACTER(LEN=LEN(group(1))) :: tgroup(SIZE(group))

  tbond(:,:)=bond(:,:)  ;  tgroup(:)=group(:)

  nca=0
  DO n=1,SIZE(group)
    IF (tgroup(n)(1:1) /= ' ') THEN
      nca=nca+1
      last=n
    ENDIF
  ENDDO
 
  IF (nca<2) RETURN
  IF (last==nca) RETURN

! erase blank lines in bond and group
! icheck is just to know if we are in an infinite loop
  icheck=0  ;  i=1

  ckloop: DO
    IF (i>nca) EXIT ckloop
    IF (tgroup(i)==' ') THEN
      DO j=i,last
        tgroup(j)=tgroup(j+1)
      ENDDO
      tgroup(last)=' '
      DO j=1,last
        DO k=i,last-1
          tbond(k,j)=tbond(k+1,j)
        ENDDO
        tbond(last,j)=0
      ENDDO
      DO j=1,last
        DO k=i,last-1
          tbond(j,k)=tbond(j,k+1)
        ENDDO
        tbond(j,last)=0
      ENDDO
      
    ENDIF
     
    IF (tgroup(i)/=' ') i=i+1
    
    icheck=icheck+1
    IF (icheck>SIZE(bond)) THEN
      WRITE(*,*) "--error-- in erase_blank"
      WRITE(*,*) "infinite loop when erasing blanks lines"
      STOP "in erase_blank"
    ENDIF
  ENDDO ckloop

  bond(:,:)=tbond(:,:)
  group(:)=tgroup(:)
      
END SUBROUTINE erase_blank

! ======================================================================
! PURPOSE: count the number of occurence of a substring in a string.
! ======================================================================
INTEGER FUNCTION countstring(line, str) RESULT(n)
  CHARACTER(LEN=*), INTENT(IN) :: line      ! input line 
  CHARACTER(LEN=*), INTENT(IN) :: str       ! string to found in line
  INTEGER :: p, ipos
 
  n=0
  IF (LEN(str) == 0) RETURN
  p = 1
  DO 
    ipos = INDEX(line(p:), str)
    IF (ipos == 0) RETURN
    n = n+1
    p = p+ipos+LEN(str)-1
  ENDDO
END FUNCTION countstring

!=======================================================================
! PURPOSE: return the rate constant at T for the arrhenius parameter
! provide as input. 
!=======================================================================
REAL FUNCTION kval(arrh,T)
  IMPLICIT NONE
  REAl,INTENT(IN) :: arrh(:)                    ! arrhenius parameter
  REAl,INTENT(IN) :: T                          ! input T
  kval=arrh(1)*(T**arrh(2))*exp(-arrh(3)/T)
  RETURN
END FUNCTION  kval


END MODULE toolbox

File: LIB/toolbox.c
/* Toolbox utility functions for GECKO-A */
#include "toolbox.h"
#include "keyparameter.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

/* Add a reference to the reference list */
void addref(const char *progname, const char *reference, int *nref, 
            char **reflist, int maxrefs, const char *chem) {
    char species[MXLFO+1];
    int i;
    
    /* Check if chem is provided */
    if (chem != NULL) {
        strncpy(species, chem, MXLFO);
        species[MXLFO] = '\0';
    } else {
        species[0] = '\0';
    }
    
    /* Check if reference already in list */
    for (i = 0; i < *nref; i++) {
        if (strcmp(reflist[i], reference) == 0) {
            return;  /* ref already in list */
        }
    }
    
    (*nref)++;
    if (*nref > maxrefs) {
        fprintf(stderr, "--error-- too many references added for the list\n");
        fprintf(stderr, "          in subroutine: %s\n", progname);
        fprintf(stderr, "size of reflist is: %d\n", maxrefs);
        if (species[0] != '\0') {
            fprintf(stderr, "          for the species: %s\n", species);
        }
        fprintf(stderr, "          reference was: %s\n", reference);
        exit(EXIT_FAILURE);
    }
    strcpy(reflist[*nref - 1], reference);
}

/* Increment number of reactions */
void addrx(const char *progname, const char *chem, int *nr, 
           int *flag, int maxflag) {
    char mesg[200];
    
    (*nr)++;
    if (*nr > maxflag) {
        snprintf(mesg, sizeof(mesg), 
                "-error- too many reactions created for species: %s in addrx", chem);
        stoperr(progname, mesg, chem);
    }
    flag[*nr - 1] = 1;
}

/* Increment number of products in a reaction */
void add1tonp(const char *progname, const char *chem, int *np) {
    (*np)++;
    if (*np > MXPD) {
        fprintf(stderr, "--error-- too many products created, np > mxpd\n");
        fprintf(stderr, "          for species: %s\n", chem);
        fprintf(stderr, "          in subroutine: %s\n", progname);
        exit(EXIT_FAILURE);
    }
}

/* Generate error message output */
void stoperr(const char *prog, const char *mesg, const char *chem) {
    const bool kill = true;  /* Set to false to continue on error */
    
    fprintf(stderr, "\n");
    fprintf(stderr, "--error-- in: %s\n", prog);
    fprintf(stderr, "%s\n", mesg);
    fprintf(stderr, "for species: %s\n", chem);
    
    if (kill) {
        exit(EXIT_FAILURE);
    }
}

/* Set bond value (symmetric) */
void setbond(int **bondtb, int x, int y, int bondvalue) {
    bondtb[x][y] = bondvalue;
    bondtb[y][x] = bondvalue;
}

/* Remove empty nodes in group and bond */
void erase_blank(int **bond, char **group, int size1, int size2) {
    int i, j, k, last, n, nca, icheck;
    int **tbond;
    char **tgroup;
    
    /* Allocate temporary arrays */
    tbond = (int **)malloc(size1 * sizeof(int *));
    for (i = 0; i < size1; i++) {
        tbond[i] = (int *)malloc(size2 * sizeof(int));
        memcpy(tbond[i], bond[i], size2 * sizeof(int));
    }
    
    tgroup = (char **)malloc(size1 * sizeof(char *));
    for (i = 0; i < size1; i++) {
        tgroup[i] = (char *)malloc(strlen(group[i]) + 1);
        strcpy(tgroup[i], group[i]);
    }
    
    /* Count non-empty groups */
    nca = 0;
    last = 0;
    for (n = 0; n < size1; n++) {
        if (tgroup[n][0] != ' ' && tgroup[n][0] != '\0') {
            nca++;
            last = n;
        }
    }
    
    if (nca < 2 || last == nca - 1) {
        /* Clean up and return */
        for (i = 0; i < size1; i++) {
            free(tbond[i]);
            free(tgroup[i]);
        }
        free(tbond);
        free(tgroup);
        return;
    }
    
    /* Erase blank lines in bond and group */
    icheck = 0;
    i = 0;
    
    while (i < nca) {
        if (tgroup[i][0] == ' ' || tgroup[i][0] == '\0') {
            /* Shift groups */
            for (j = i; j < last; j++) {
                strcpy(tgroup[j], tgroup[j+1]);
            }
            tgroup[last][0] = '\0';
            
            /* Shift bond rows */
            for (j = 0; j <= last; j++) {
                for (k = i; k < last; k++) {
                    tbond[k][j] = tbond[k+1][j];
                }
                tbond[last][j] = 0;
            }
            
            /* Shift bond columns */
            for (j = 0; j <= last; j++) {
                for (k = i; k < last; k++) {
                    tbond[j][k] = tbond[j][k+1];
                }
                tbond[j][last] = 0;
            }
        } else {
            i++;
        }
        
        icheck++;
        if (icheck > size1) {
            fprintf(stderr, "--error-- in erase_blank\n");
            fprintf(stderr, "infinite loop when erasing blank lines\n");
            exit(EXIT_FAILURE);
        }
    }
    
    /* Copy back results */
    for (i = 0; i < size1; i++) {
        memcpy(bond[i], tbond[i], size2 * sizeof(int));
        strcpy(group[i], tgroup[i]);
    }
    
    /* Clean up */
    for (i = 0; i < size1; i++) {
        free(tbond[i]);
        free(tgroup[i]);
    }
    free(tbond);
    free(tgroup);
}

/* Count the number of occurrences of a substring in a string */
int countstring(const char *line, const char *str) {
    int n = 0;
    const char *p;
    size_t len;
    
    len = strlen(str);
    if (len == 0) return 0;
    
    p = line;
    while ((p = strstr(p, str)) != NULL) {
        n++;
        p += len;
    }
    
    return n;
}

/* Return the rate constant at T for the Arrhenius parameters */
double kval(const double *arrh, double T) {
    /* arrh[0] = A, arrh[1] = n, arrh[2] = Ea/R */
    return arrh[0] * pow(T, arrh[1]) * exp(-arrh[2] / T);
}
